src/config.rs

```diff
@@ -131,38 +122,145 @@ pub struct Opt {
+        long = "source_uri_base",
+        value_name = "alternate fetch base uri or dir for notification uri"
+    )]
+    pub source_uri_base: Option<FetchSource>,
+
+    #[structopt(skip)]
+    pub fetch_map: Option<FetchMap>,
 }
 
-fn log_without_target(
-    out: fern::FormatCallback,
-    message: &std::fmt::Arguments,
-    record: &log::Record,
-) {
-    out.finish(format_args!(
-        "{} {}: {}",
-        chrono::Local::now().format("%Y/%m/%d %H:%M:%S"),
-        record.level(),
-        message,
-    ))
+impl Config {
+    pub fn rsync_enabled(&self) -> bool {
+        !self.rsync_disable
+    }
+
+    pub fn rsync_dir_use_symlinks(&self) -> bool {
+        if cfg!(unix) {
+            !self.rsync_dir_force_moves
+        } else {
+            false
+        }
+    }
+
+    pub fn rsync_dir_current(&self) -> PathBuf {
+        self.rsync_dir.join("current")
+    }
+
+    pub fn fetcher(&self) -> Fetcher {
+        Fetcher::new(self.notification_uri.clone(), self.fetch_map.clone())
+    }
+
+    pub fn state_path(&self) -> PathBuf {
+        self.state_dir.join("rrdp-state.json")
+    }
+
+    pub fn rsync_state_path(&self) -> PathBuf {
+        self.state_dir.join("rsync-state.json")
+    }
 }
 
-fn log_with_target(out: fern::FormatCallback, message: &std::fmt::Arguments, record: &log::Record) {
-    out.finish(format_args!(
-        "{} {} [{}] {}",
-        chrono::Local::now().format("%Y/%m/%d %H:%M:%S"),
-        record.level(),
-        record.target(),
-        message,
-    ))
+pub fn configure() -> Result<Config> {
+    let config = Config::from_args();
+    post_configure(config)
+}
+
+#[cfg(test)]
+pub fn create_test_config(
+    work_dir: &Path,
+    notification_uri: Https,
+    source_uri_base: &str,
+) -> Config {
+    let source_uri_base = FetchSource::File(PathBuf::from(source_uri_base));
+
+    let state_dir = work_dir.join("state");
+    let rrdp_dir = work_dir.join("rrdp");
+    let rsync_dir = work_dir.join("rsync");
+    let pid_file = work_dir.join("krill-sync.pid");
+
+    let config = Config {
+        verbose: 0,
+        quiet: false,
+        force_snapshot: false,
+        force_update: false,
+        pid_file,
+        state_dir,
+        rrdp_dir,
+        rrdp_notify_delay: 0,
+        rsync_dir,
+        rsync_dir_force_moves: false,
+        rsync_disable: false,
+        cleanup_after: 2,
+        insecure: false,
+        notification_uri,
+        source_uri_base: Some(source_uri_base),
+        fetch_map: None, // will be set in post_configure
+    };
+    post_configure(config).unwrap()
 }
 
-pub fn configure() -> Result<Opt> {
-    let mut opt = Opt::from_args();
+pub fn post_configure(mut config: Config) -> Result<Config> {
+    initialize_logging(&config);
+
+    let base_uri = config
+        .notification_uri
+        .parent()
+        .ok_or_else(|| anyhow!("Notification URI should contain a path to a file"))?;
+
+    if let Some(base_fetch) = config.source_uri_base.as_ref().cloned() {
```

Comment:
Minor nit: Replace `.cloned()` here with `FetchMap::new(base_uri, base_fetch.clone())` below, i.e. don't clone if not needed.

Reply:
ack